packages/commerce-sdk-react/src/hooks/types.ts

```diff
@@ -38,35 +50,91 @@ export interface ApiClients {
     shopperSearch: ShopperSearch<ApiClientConfigParams>
 }
 
+export type ApiClient = ApiClients[keyof ApiClients]
+
+// --- API HELPERS --- //
+
+/**
+ * Generic signature of the options objects used by commerce-sdk-isomorphic
+ */
+export type ApiOptions<
+    Parameters extends object = Record<string, unknown>,
+    Headers extends Record<string, string> = Record<string, string>,
+    Body extends object | unknown[] | undefined = Record<string, unknown> | unknown[] | undefined
+> = {
+    parameters?: Parameters
+    headers?: Headers
+    body?: Body
+}
+
+/**
+ * Generic signature of API methods exported by commerce-sdk-isomorphic
+ */
+export type ApiMethod<Options extends ApiOptions, Data> = {
+    (options: Options): Promise<Data>
+}
+
 /**
  * The first argument of a function.
  */
-export type Argument<T extends (arg: any) => unknown> = Parameters<T>[0]
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+export type Argument<T extends (arg: any) => unknown> = NonNullable<Parameters<T>[0]>
 
 /**
  * The data type returned by a commerce-sdk-isomorphic method when the raw response
  * flag is not set.
  */
-export type DataType<T extends (arg: any) => Promise<unknown>> = T extends (
-    arg: any
-) => Promise<Response | infer R>
-    ? R
-    : never
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+export type DataType<T> = T extends ApiMethod<any, Response | infer R> ? R : never
 
 /**
- * Modified version of React Query's Mutation Function. Added a second argument
- * API clients.
+ * Merged headers and parameters from client config and options, mimicking the behavior
+ * of commerce-sdk-isomorphic.
  */
-export type IMutationFunction<TData = unknown, TVariables = unknown> = (
-    variables: TVariables,
-    apiClients: ApiClients
-) => Promise<TData>
+export type MergedOptions<Client extends ApiClient, Options extends ApiOptions> = RequireKeys<
+    ApiOptions<
+        NonNullable<Client['clientConfig']['parameters'] & Options['parameters']>,
+        NonNullable<Client['clientConfig']['headers'] & Options['headers']>,
+        // `body` may not exist on `Options`, in which case it is `unknown` here. Due to the type
+        // constraint in `ApiOptions`, that is not a valid value. We must replace it with `never`
+        // to indicate that the result type does not have a `body`.
+        unknown extends Options['body'] ? never : Options['body']
+    >,
+    'parameters' | 'headers'
+>
 
-/**
- * Modified version of React Query's Query Function. Added a second argument
- * API clients.
- */
-export type IQueryFunction<TData = unknown> = (
-    context: QueryFunctionContext<QueryKey>,
-    apiClients: ApiClients
-) => Promise<TData>
+// --- CACHE HELPERS --- //
+
+export type ApiQueryKey =
+    // | readonly string[] // TODO: Is this needed?
+    readonly [...path: string[], parameters: Record<string, unknown>]
```

Comment:
The format of the query key will likely change. I think for our composable-predicate approach, the best solution would be an array with a single object. Something along the lines of `[{ path, options, ... }]`.